File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int n = matrix.length;
        int[][] dp = new int[n][n];
        
        for (int i = 0; i < n; i++) {
            dp[0][i] = matrix[0][i];
        }
        
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (0 < j && j < n-1) {
                    dp[i][j] = min(dp[i-1][j-1], dp[i-1][j], dp[i-1][j+1]) + matrix[i][j];
                } else if (j == 0) {
                    dp[i][j] = min(dp[i-1][j], dp[i-1][j+1]) + matrix[i][j];
                } else {
                    dp[i][j] = min(dp[i-1][j], dp[i-1][j-1]) + matrix[i][j];
                }
            }
        }
        
        
        return min(dp[n-1]);
    }
    
    
    private int min(int x, int y) {
        return Math.min(x, y);
    }
    
    private int min(int x, int y, int z) {
        return Math.min(Math.min(x, y), z);
    }
    
    private int min(int[] arr) {
        int res = arr[0];
        for (int num : arr) {
            res = Math.min(res, num);
        }
        return res;
    }
}

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.rs
impl Solution {
    pub fn min_falling_path_sum(matrix: Vec<Vec<i32>>) -> i32 {
        let n : usize = matrix.len();
        let m : usize = matrix[0].len();

        let mut dp : Vec<Vec<i32>> = vec![vec![1_000_000; m]; n];
        for j in 0..m {
            dp[0][j] = matrix[0][j];
        }
        for i in 1..n {
            for j in 0..m {
                if j != 0 {
                    dp[i][j] = dp[i][j].min(dp[i-1][j-1] + matrix[i][j]);
                }
                if j != m-1 {
                    dp[i][j] = dp[i][j].min(dp[i-1][j+1] + matrix[i][j]);
                }
                dp[i][j] = dp[i][j].min(dp[i-1][j] + matrix[i][j]);
            }
        }
        *dp[n-1].iter().min().unwrap()
    }
}
